# Project Euler

The benefits of working on the Project Euler problems are too great to ignore:

- They’ll expose whether you understand the basic operations of the language you’re using, and they can be solved in any programming language
- You’ll learn how to write solid, standard, vanilla code. No libraries needed, so you won’t have crutches to rely on
- You’ll have to think about how you store and access data in memory
- You’ll need to optimize processing speed and think hard about algorithms

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

#!pip install jupyternotify
import jupyternotify
ip = get_ipython()
ip.register_magics(jupyternotify.JupyterNotifyMagics)
## Run %%notify to create notification for completed cell
%notify

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Problem 1: Multiples of 3 and 5

#### If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

* Find the sum of all the multiples of 3 or 5 below 1000

In [41]:
nums = list(np.linspace(1, 9, 9, dtype=int))
nums

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [44]:
numbers = []
for num in nums:
    if (num % 3) == 0:
        numbers.append(num)
    elif (num % 5) == 0:
        numbers.append(num)
print(numbers, "are all multipes of 3 or 5")
tot = 0
for num in numbers:
    tot += num
print(tot, "is the sum of these multiples")

[3, 5, 6, 9] are all multipes of 3 or 5
23 is the sum of these multiples


In [46]:
nums = list((range(1000)))
numbers = []
tot = 0
for num in nums:
    if (num % 3) == 0:
        numbers.append(num)
    elif (num % 5) == 0:
        numbers.append(num)
print("There are {} multiples of 3 or 5 below 1000".format(len(numbers)))
for num in numbers:
    tot += num
print(tot, "is the sum of these multiples")

There are 467 multiples of 3 or 5 below 1000
233168 is the sum of these multiples


## SOLUTION: Problem 1: 
**233,168 is the sum of all the multiples of 3 or 5 below 1000**

## Problem 2: Even Fibonacci Numbers

### Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

**1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...**

- By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [82]:
fibonacci = [1, 2]
while fibonacci[-1] <= 4000000:
    new2 = fibonacci[-2]
    new1 = fibonacci[-1]
    fibonacci.append(new1 + new2)
tot = 0
for fib in fibonacci[:-1]:
    if fib % 2 == 0:
        tot += fib
        print(fib, "added to total")
print()
print(tot, "is the sum of all even-valued terms less than 4 million")

2 added to total
8 added to total
34 added to total
144 added to total
610 added to total
2584 added to total
10946 added to total
46368 added to total
196418 added to total
832040 added to total
3524578 added to total

4613732 is the sum of all even-valued terms less than 4 million


## SOLUTION: Problem 2: 
**4,613,732 is the sum of all even-valued terms in the Fibonacci Sequence that are less than 4 Million**

## Problem 3: Largest Prime Factor

### The prime factors of 13,195 are 5, 7, 13 and 29.

* What is the largest prime factor of the number 600,851,475,143 ?

In [113]:
# Initialize a list
primes = []
for possiblePrime in range(2, 100851475143):
    # Assume number is prime until shown it is not. 
    isPrime = True
    for num in range(2, possiblePrime):
        if possiblePrime % num == 0:
            isPrime = False
    if isPrime:
        primes.append(possiblePrime)

KeyboardInterrupt: 

In [ ]:
print(primes[-1], "is the largest prime factor of the number 600,851,475,143")
%notify